### PR TITLE
parquet2 implementation backed by parquet2 feature gate

### PR DESCRIPTION
**Pull request by houqp**
decouple core from arrow

# Description

WIP parquet2 implementation. The goal of this PR is to implement full read support leveraging parquet2. Write support is out of the scope and should be added as follow up PRs. Arrow2 integration is also out of the scope and should be added through follow up PR.

Currently all read tests are passing:

```
cargo test --no-default-features --features=arrow2,parquet2 
```

A quick benchmark shows more than 50% performance boost on checkpoint deserialization. Only tested with a very tiny checkpoint from the golden dataset, I would expect the performance gap would be bigger for larger real world tables.

Todo:

- [x] clean up duplicated code
- [x] support parsing map type
- [x] support parsing list type
- [x] benchmark

# Related Issue(s)

blocks #310 

# Documentation

### CODE REVIEW

**Comment by houqp** [comment]
Sounds good @ritchie46 , I will complete the parquet parsing support for map and list this weekend. But the branch I have here right now should be enough to unblock ploars integration.

**Comment by ahmedriza** [comment]
There are currently some issues in `parquet2` with regard to nested types.  See https://github.com/jorgecarleitao/arrow2/issues/937 for example.

**Comment by andrei-ionescu** [comment]
@houqp, any updates on this?

**Comment by houqp** [comment]
@andrei-ionescu I have implemented all the data types other than map and nested list, so it's very close to be complete. However, my time is limited now, so progress will be slow. Anyone if welcome to collaborate on this branch to push this over the finish line :)

**Comment by houqp** [comment]
alright, this branch is now feature complete ;) now it's time to catch up to latest delta-rs and arrow2/parquet2 releases.

File: rust/tests/dataframe_test.rs [hunk 1]
```diff
@@ -1,89 +1,83 @@
-extern crate arrow;
-extern crate deltalake;
 #[cfg(feature = "rust-dataframe-ext")]
-extern crate rust_dataframe;
+mod rust_dataframe {
```
Comment (houqp) [inline rust/tests/dataframe_test.rs:2]: no change in test, just moved code into its own mod to make conditional compilation annotation cleaner.

**Comment by houqp** [comment]
@wjones127 @roeap ready for review.

File: rust/src/action/parquet2_read/dictionary/binary.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+use parquet2::encoding::get_length;
+use parquet2::error::Error;
+
+// FIXME: replace vec with reference
```
Comment (roeap) [inline rust/src/action/parquet2_read/dictionary/binary.rs:4]: should this be fixed in this PR?

File: rust/src/action/parquet2_read/map.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
+// use std::collections::HashMap;
```
Comment (roeap) [inline rust/src/action/parquet2_read/map.rs:1]: remove?

File: rust/src/action/parquet2_read/mod.rs [hunk 4]
```diff
@@ -0,0 +1,864 @@
...
+        let schema_path = &column_desc.path_in_schema;
+
+        let deserialize_column_page = match schema_path[0].as_ref() {
+            "txn" => deserialize_txn_column_page,
+            "add" => deserialize_add_column_page,
+            "remove" => deserialize_remove_column_page,
+            "metaData" => deserialize_metadata_column_page,
+            "protocol" => deserialize_protocol_column_page,
+            "commitInfo" => deserialize_commit_info_column_page,
+            "cdc" => deserialize_cdc_column_page,
```
Comment (roeap) [inline rust/src/action/parquet2_read/mod.rs:665]: ❤️

File: rust/tests/dataframe_test.rs (hunk 1 above)
Comment (roeap) [reply rust/tests/dataframe_test.rs:2]: don't think it matters much but `#![cfg(feature = "rust-dataframe-ext")]` might also work?

As a side note I was wondering if this still needs to / should be supported? It seems the repository for rust_dataframe has been archived, or did just move somewhere else?

**Review by roeap — COMMENTED** [review]
Only had some very minor comments. While not an expert on parquet reading, I felt the tests should cover that quite nicely!

The one thing I was wondering is the `impl`'s for the actions, or the action.rs file in general. It seems we have alot of feature flags there now, and maybe it would be cleaner to have the parquet specific stuff also moved into its own mod. Maybe adopt something ile the trait based approach in the new parquet2 implementation?  

As this also seems like a step towards supporting `polars`, we will face the question of having good integration points to support various backends / engines or however one wnats to call it :).

great work!

File: rust/src/action/parquet2_read/dictionary/binary.rs (hunk 2 above)
Comment (houqp) [reply rust/src/action/parquet2_read/dictionary/binary.rs]: good catch, totally forgot about this one.

**Comment by houqp** [comment]
> maybe it would be cleaner to have the parquet specific stuff also moved into its own mod

Good idea, I have moved all those code into a `parquet_read` mod to keep action lean.

I chatted this with Andy and Jorge at the Data+AI summit a couple of months ago, the long term goal is to develop an Arrow trait that allows users to switch between arrow-rs and arrow2 in different projects including datafusion and delta-rs. This will also open up the possibility for a 3rd GPU based arrow implementation ;)

**Review by roeap — APPROVED** [review]
LGTM!

**Pull request merged**